File: app/src/main/java/com/bogdanorzea/popularmovies/ui/main/PosterAdapter.java
package com.bogdanorzea.popularmovies.ui.main;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bogdanorzea.popularmovies.R;
import com.bogdanorzea.popularmovies.model.object.Movie;
import com.bogdanorzea.popularmovies.ui.details.DetailsActivity;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class PosterAdapter extends RecyclerView.Adapter<PosterAdapter.ViewHolder> {
    private final Context context;
    private List<Movie> mMovies;

    public PosterAdapter(Context context) {
        this.context = context;
        mMovies = new ArrayList<>();
    }

    public List<Movie> getMovies() {
        return mMovies;
    }

    public boolean isEmpty() {
        return mMovies.size() < 1;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View linearLayout = LayoutInflater.from(context).inflate(R.layout.item_poster, parent, false);

        return new ViewHolder(linearLayout);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Movie movie = mMovies.get(position);

        holder.itemView.setTag(movie);

        Picasso.with(context)
                .load(movie.getPosterUrl())
                .error(R.drawable.missing_cover)
                .into(holder.coverImage);
    }

    @Override
    public int getItemCount() {
        return mMovies.size();
    }

    public void addMovies(@NonNull List<Movie> data) {
        int size = getItemCount();
        mMovies.addAll(data);
        if (size > 0) {
            notifyItemRangeInserted(size, data.size());
        } else {
            notifyDataSetChanged();
        }
    }

    static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        ImageView coverImage;

        ViewHolder(View view) {
            super(view);

            coverImage = view.findViewById(R.id.poster);

            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            Context context = view.getContext();

            Movie currentMovie = (Movie) itemView.getTag();

            Intent intent = new Intent(context, DetailsActivity.class);
            intent.putExtra(DetailsActivity.MOVIE_ID, currentMovie.id);
            context.startActivity(intent);
        }
    }

}

File: app/src/main/java/com/bogdanorzea/popularmovies/ui/main/PosterAdapter.kt
package com.bogdanorzea.popularmovies.ui.main

import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.bogdanorzea.popularmovies.R
import com.bogdanorzea.popularmovies.model.`object`.Movie
import com.bogdanorzea.popularmovies.ui.details.DetailsActivity
import com.squareup.picasso.Picasso
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_poster.*

class PosterAdapter(private val context: Context) : RecyclerView.Adapter<PosterAdapter.ViewHolder>() {
    val movies: MutableList<Movie> = mutableListOf()

    fun isEmpty(): Boolean = movies.size < 1

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layout = LayoutInflater.from(context).inflate(R.layout.item_poster, parent, false)

        return ViewHolder(layout)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bindMovie(movies[position])
    }

    override fun getItemCount(): Int = movies.size

    fun addMovies(data: List<Movie>) {
        val size = itemCount
        movies.addAll(data)
        if (size > 0) {
            notifyItemRangeInserted(size, data.size)
        } else {
            notifyDataSetChanged()
        }
    }

    class ViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView), LayoutContainer, View.OnClickListener {

        fun bindMovie(movie: Movie) {
            Picasso.with(containerView.context)
                    .load(movie.getPosterUrl())
                    .error(R.drawable.missing_cover)
                    .into(poster)

            containerView.tag = movie
            containerView.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            val context = view.context

            val (_, _, _, _, id) = itemView.tag as Movie

            val intent = Intent(context, DetailsActivity::class.java)
            intent.putExtra(DetailsActivity.MOVIE_ID, id)
            context.startActivity(intent)
        }
    }

}
